File: engine-rust/src/ai/simple.rs
use crate::core::battle::{is_battle_over, step_battle, BattleOptions};
use crate::core::state::{Action, ActionType, BattleState};
use crate::core::utils::get_active_creature;
use crate::data::moves::MoveDatabase;

pub fn choose_highest_power(state: &BattleState, player_id: &str) -> Option<Action> {
    let player = state.players.iter().find(|p| p.id == player_id)?;
    let active = get_active_creature(state, player_id)?;
    if active.hp <= 0 {
        return None;
    }

    let target_id = state
        .players
        .iter()
        .find(|p| p.id != player_id)
        .map(|p| p.id.clone())?;

    let move_db = MoveDatabase::minimal();
    let mut best_move_id = active.moves.first()?.clone();
    let mut best_power = -1;

    for move_id in &active.moves {
        if let Some(move_data) = move_db.get(move_id) {
            let power = move_data.power.unwrap_or(0);
            if power > best_power {
                best_power = power;
                best_move_id = move_id.clone();
            }
        }
    }

    Some(Action {
        player_id: player.id.clone(),
        action_type: ActionType::Move,
        move_id: Some(best_move_id),
        target_id: Some(target_id),
        slot: None,
        priority: None,
    })
}

pub fn run_auto_battle(
    state: &BattleState,
    rng: &mut dyn FnMut() -> f64,
    chooser: fn(&BattleState, &str) -> Option<Action>,
) -> BattleState {
    let mut next = state.clone();
    let mut turns = 0;
    while !is_battle_over(&next) && turns < 100 {
        turns += 1;
        let mut actions = Vec::new();
        for player in &next.players {
            if let Some(action) = chooser(&next, &player.id) {
                actions.push(action);
            }
        }
        if actions.is_empty() {
            break;
        }
        next = step_battle(&next, &actions, rng, BattleOptions::default());
    }
    next
}

File: engine-legacy/ai/simple.js
const { stepBattle, isBattleOver } = require("../core/battle");
const { moves } = require("../data/moves");
const { getActiveCreature } = require("../core/utils");

/**
 * Choose the move with highest nominal power for the active creature.
 * @param {import("../core/state").BattleState} state
 * @param {string} playerId
 * @returns {{playerId: string, moveId: string, targetId?: string}|null}
 */
function chooseHighestPower(state, playerId) {
  const player = state.players.find((p) => p.id === playerId);
  const active = getActiveCreature(state, playerId);
  if (!player) return null;

  // If active pokemon is fainted or needs to switch, choose a random alive backup
  if (!active || active.hp <= 0 || active.statuses.some(s => s.id === "pending_switch")) {
      const backupIndex = player.team.findIndex((p, i) => p.hp > 0 && i !== player.activeSlot);
      if (backupIndex >= 0) {
          return { type: "switch", playerId, slot: backupIndex };
      }
      if (!active || active.hp <= 0) return { type: "wait", playerId };
  }

  const targetId = state.players.find((p) => p.id !== playerId)?.id;
  if (!targetId) return null;

  let bestMoveId = active.moves[0];
  let bestPower = -Infinity;
  for (const moveId of active.moves) {
    const move = moves[moveId];
    if (!move) continue;
    const power = (move.effects || [])
      .filter((e) => e.type === "damage")
      .reduce((max, e) => Math.max(max, e.power ?? 0), -Infinity);
    if (power > bestPower) {
      bestPower = power;
      bestMoveId = moveId;
    }
  }
  return { type: "move", playerId, moveId: bestMoveId, targetId };
}

/**
 * Run an auto battle until completion.
 * @param {import("../core/state").BattleState} state
 * @param {() => number} rng
 * @param {(state: import("../core/state").BattleState, playerId: string) => Object} chooser
 * @returns {import("../core/state").BattleState}
 */
function runAutoBattle(state, rng = Math.random, chooser = chooseHighestPower) {
  let next = state;
  let turns = 0;
  while (!isBattleOver(next) && turns < 100) {
    turns++;
    const actions = [];
    for (const player of next.players) {
      const action = chooser(next, player.id);
      if (action) actions.push(action);
    }
    if (actions.length === 0) break;
    next = stepBattle(next, actions, rng);
  }
  return next;
}

module.exports = { chooseHighestPower, runAutoBattle };
